app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodModal.tsx

```diff
@@ -0,0 +1,173 @@
+              <Elements stripe={stripePromise} options={options}>
+                <SetupStripePaymentMethodForm
+                  disclosure={disclosure}
+                  clientSecret={clientSecret}
+                  successCallback={successCallback}
+                />
+              </Elements>
+            </VStack>
+          </Box>
+        </ModalBody>
+      </ModalContent>
+    </Modal>
+  );
+}
+
+function SetupStripePaymentMethodForm({
+  disclosure,
+  clientSecret,
+  successCallback,
+}: {
+  disclosure: UseDisclosureReturn;
+  clientSecret: string;
+  successCallback?: () => void;
+}) {
+  const stripe = useStripe();
+  const elements = useElements();
+  const selectedProject = useSelectedProject().data;
+  const utils = api.useContext();
+
+  // Code from stripe docs
+  const [handleAddPaymentDetail, addPaymentDetailLoading] = useHandledAsyncCallback(async () => {
+    const setupIntent = await addPaymentDetail();
+
+    if (setupIntent) {
+      switch (setupIntent.status) {
+        case "succeeded":
+          if (successCallback) {
+            successCallback();
+          } else {
+            toast({
+              description: "Your payment method has been saved!",
+              status: "success",
+              duration: 5000,
+              isClosable: true,
+            });
+          }
+
+          break;
+
+        case "processing":
+          toast({
+            description: "Processing payment details. Please wait a few minutes.",
+            status: "success",
+            duration: 5000,
+            isClosable: true,
+          });
+          break;
+
+        case "requires_payment_method":
+          toast({
+            description: "Failed to process payment details. Please try another payment method.",
+            status: "error",
+            duration: 5000,
+            isClosable: true,
+          });
+          break;
+      }
+      disclosure.onClose();
+    }
+
+    await utils.payments.getPaymentMethods.invalidate();
+  }, [utils, addPaymentDetail]);
+
+  async function addPaymentDetail() {
+    if (!stripe || !elements) {
+      return;
+    }
+
+    const { error } = await stripe.confirmSetup({
+      elements,
+      confirmParams: {
+        return_url: `${env.NEXT_PUBLIC_HOST}/p/${
+          selectedProject?.slug ?? ""
+        }/billing/payment-methods`,
+      },
+      redirect: "if_required",
+    });
+
+    if (error) {
+      toast({
+        description: error.message,
+        status: "error",
+        duration: 5000,
+        isClosable: true,
+      });
+    } else {
+      const { setupIntent } = await stripe.retrieveSetupIntent(clientSecret);
+
+      return setupIntent;
+    }
+  }
+
+  return (
+    <Box style={{ width: "100%" }}>
+      <PaymentElement />
+      <HStack w="full" justifyContent="end">
+        <Button colorScheme="gray" minW={24} onClick={() => handleAddPaymentDetail()}>
+          {addPaymentDetailLoading ? <Spinner /> : "Save"}
```

Comment:
Is this the button that I mentioned should be "Save and Pay" if you're paying right after?

Reply:
Done!
I changed a toast keeping the button to make it simpler. It makes sense to change the button as well.